QccPack-0.61-1/src/libQccPackSQ/utils/man1/squniform.1
.TH SQUNIFORM 1 "QCCPACK" ""
.SH NAME
squniform \- make uniform scalar quantizer
.SH SYNOPSIS
.B squniform
.RB "[\|" \-max
.IR  max\_value "\|]"
.RB "[\|" \-min
.IR min\_value "\|]"
.RB "[\|" \-u 
.IR u "\|]"
.RB "[\|" \-A
.IR A "\|]"
.RB "[\|" \-nl
.IR num\_levels "\|]"
.RB "[\|" \-q
.IR stepsize "\|]"
.RB "[\|" \-d
.IR deadzone "\|]"
.I quantizer\_file
.SH OPTIONS
.TP
.BI \-max " max\_value"
Float. Maximum value of the range of the quantizer.  Default value = 255.
.TP
.BI \-min " min\_value"
Float. Minimum value of the range of the quantizer.  Default value = 0.
.TP
.BI \-u " u"
Float.  Apply a u-law compander to the uniform quantizer.
.TP
.BI \-A " A"
Float.  Apply a A-law compander to the uniform quantizer.
.TP
.BI \-nl " num\_levels"
Integer.  Number of quantizer intervals. Default = 255.
.TP
.BI \-q " stepsize"
Float. Stepsize of uniform quantizer.
.TP
.BI \-d " deadzone"
Float. Deadzone of quantizer.
.SH DESCRIPTION
.LP
.B squniform
designs scalar quantizers based on the uniform scalar quantizer.
.B squniform
supports simple uniform scalar quantization, uniform scalar quantization
with a dead zone about zero, and u-law and A-law companding.
.LP
A uniform scalar quantizer can be created in two ways.
The first way is to specify the scalar quantizer stepsize with
the 
.B \-q
option; this type of quantizer will be called hereafter the "simple"
uniform scalar quantizer.
Encoding using this simple quantizer (i.e., with
.BR sqencode (1))
is merely a division by the stepsize; i.e.,
.IR index =
.BR rint (
.IR value / stepsize ).
.LP
The second form of uniform quantizer supported by
.B squniform
is a "general" uniform scalar quantizer.
In this type of quantizer, the reproduction levels and boundaries
are explicitly given in the output file.
The range of the general quantizer is 
.I min\_value 
through
.I max\_value 
and the number of quantization levels is
.IR num\_levels .
.B squniform
constructs
.I num\_levels
equally-sized quantizer bins between
.I min\_value
and
.IR max\_value .
Note: the key difference between the simple and general uniform
scalar quantizers is
that the range of the simple quantizer is implied to
be infinite while the range of the general scalar quantizer is finite.
However, the first and last quantizer bins of the general quantizer are
considered to be "overload" regions.
Note: quantizing with the general uniform scalar quantizer (via, for
example,
.BR sqencode (1))
requires a nearest-neighbor search and is thus slower than using a
simple quantizer.
.LP
A dead-zone quantizer can be constructed by giving both the
.I stepsize
of the quantizer (via the
.B \-q
option) and the
.I deadzone
bin size (via the
.B \-d
option).
For a dead-zone quantizer, 
.I deadzone
is the size of the quantizer bin centered at zero; that is, all values
between
.I -deadzone
and 
.I deadzone
are quantized to zero.
Beyond this dead zone about zero, the rest of the quantizer is uniform, with
bins of size
.IR stepsize .
Note: the simple uniform scalar quantizer can be considered to be a
deadzone quantizer with
.IR deadzone = stepsize .
.LP
Companding may be applied to the general uniform scalar quantizer to produce
u-law and A-law scalar quantizers.
With the
.B \-u
or
.B \-A
options, logarithmic companders with u-law or A-law
characteristics, respectively, are applied to the general
quantizer before it is output (in effect making the quantizer nonuniform).
One of u-law or A-law, but not both, may be specified.
The value
.I u
must be nonnegative;
.I A
must be 1.0 or greater.
Options
.B \-u
and
.B \-A
may not be used in conjunction with
the
.B \-q
option.
.SH "SEE ALSO"
.BR sqlloyd (1),
.BR sqencode (1),
.BR QccPackSQ (3),
.BR QccPack (3)

A. Gersho and R. Gray, 
.IR "Vector Quantization and Signal Compression" .
Norwell, MA: Kluwer Academic Publishers, 1992, pp. 151-161.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackSQ/utils/squniform.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "squniform.h"

#define USG_STRING "[-max %f:max_value] [-min %f:min_value] [-u %: %f:u] [-A %: %f:A] [-nl %d:num_levels] [-q %f:stepsize] [-d %f:deadzone] %s:quantizer_file"

QccSQScalarQuantizer Quantizer;

float MaxValue = 255;
float MinValue = 0;

int ULaw = 0;
float U;

int ALaw = 0;
float A;

int NumLevels = 255;

float StepSize = -1;
float DeadZone = -1;

int main(int argc, char *argv[])
{

  QccInit(argc, argv);

  QccSQScalarQuantizerInitialize(&Quantizer);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &MaxValue,
                         &MinValue,
                         &ULaw,
                         &U,
                         &ALaw,
                         &A,
                         &NumLevels,
                         &StepSize,
                         &DeadZone,
                         Quantizer.filename))
    QccErrorExit();
      
  if ((DeadZone > 0) && (StepSize <= 0))
    {
      QccErrorAddMessage("%s: Must define stepsize > 0 for dead-zone quantizer",
                         argv[0]);
      QccErrorExit();
    }

  if ((ALaw || ULaw) && (StepSize > 0))
    {
      QccErrorAddMessage("%s: Cannot specify both companding and uniform-quantizer stepsize",
                         argv[0]);
      QccErrorExit();
    }

  if (StepSize > 0)
    {
      if (DeadZone >= 0)
        {
          Quantizer.type = QCCSQSCALARQUANTIZER_DEADZONE;
          Quantizer.deadzone = DeadZone;
          Quantizer.stepsize = StepSize;
        }
      else
        {
          Quantizer.type = QCCSQSCALARQUANTIZER_UNIFORM;
          Quantizer.stepsize = StepSize;
        }
    }
  else
    {
      Quantizer.type = QCCSQSCALARQUANTIZER_GENERAL;

      if (ULaw && ALaw)
        {
          QccErrorAddMessage("%s: Cannot specify both u-Law and A-Law companding",
                             argv[0]);
          QccErrorExit();
        }
      
      if (NumLevels < 1)
        {
          QccErrorAddMessage("%s: Number of levels must be 1 or greater",
                             argv[0]);
          QccErrorExit();
        }
      Quantizer.num_levels = NumLevels;
      
      if (MaxValue <= MinValue)
        {
          QccErrorAddMessage("%s: MaxValue must be greater than MinValue",
                             argv[0]);
          QccErrorExit();
        }
      
      if (ULaw)
        {
          if (QccSQULawMakeQuantizer(&Quantizer,
                                     (double)U,
                                     (double)MaxValue, (double)MinValue))
            {
              QccErrorAddMessage("%s: Error calling QccSQULawMakeQuantizer()",
                                 argv[0]);
              QccErrorExit();
            }
        }
      else if (ALaw)
        {
          if (QccSQALawMakeQuantizer(&Quantizer,
                                     (double)A,
                                     (double)MaxValue, (double)MinValue))
            {
              QccErrorAddMessage("%s: Error calling QccSQALawMakeQuantizer()",
                                 argv[0]);
              QccErrorExit();
            }
        }
      else
        {
          if (QccSQUniformMakeQuantizer(&Quantizer,
                                        (double)MaxValue, (double)MinValue,
                                        QCCSQ_NOOVERLOAD))
            {
              QccErrorAddMessage("%s: Error calling QccSQUniformMakeQuantizer()",
                                 argv[0]);
              QccErrorExit();
            }
        }
    }

  if (QccSQScalarQuantizerWrite(&Quantizer))
    {
      QccErrorAddMessage("%s: Error calling QccSQScalarQuantizerWrite()",
                         argv[0]);
      QccErrorExit();
    }

  QccSQScalarQuantizerFree(&Quantizer);

  QccExit;
}
